src/Dictionaries/RegexpTreeDictionary.cpp

```diff
@@ -0,0 +1,398 @@
+    
+    auto type_call = [&](const auto & dictionary_attribute_type)
+    {
+	using Type = std::decay_t<decltype(dictionary_attribute_type)>;
+        using AttributeType = typename Type::AttributeType;
+        using ValueType = DictionaryValueType<AttributeType>;
+        
+	if constexpr (std::is_same_v<ValueType, StringRef>)
+            attribute.string_arena = std::make_unique<Arena>();
+        
+	attribute.container.emplace<ContainerType<ValueType>>(configuration.initial_array_size, ValueType());
+    };
+    
+    callOnDictionaryAttributeType(dictionary_attribute.underlying_type, type_call);
+    
+    return attribute;
+}
+
+void RegexpTreeDictionary::createAttributes()  
+{
+    const auto size = dict_struct.attributes.size();
+    attributes.reserve(size);
+
+    for (const auto & attribute : dict_struct.attributes)
+    {
+	if (attribute.is_nullable)
+                throw Exception(ErrorCodes::UNSUPPORTED_METHOD,
+                    "Array or nullable attributes not supported for dictionary of type {}",
+                    getTypeName());
+
+        if (attribute.hierarchical)
+            throw Exception(ErrorCodes::BAD_ARGUMENTS, "Hierarchical attributes not supported by {} dictionary.",
+                            getDictionaryID().getNameForLogs());
+	
+	attributes.push_back(createAttribute(attribute));
+    }
+}
+
+template <typename ValueSetter>
+void func(
+	RegexpTreeNode node,
+	String regexp_,
+	bool isLast,
+	size_t attribute_index,
+	size_t rows,
+	std::vector<StringRef> user_agents,
+	ValueSetter && set_value,
+	size_t keys_found)
+{
+    bool isRegexp = false;
+    regexp_+=node.regexp;
+    String attribute = node.attributes[attribute_index];
+    String final_attribute;
+    
+    if(startsWith(attribute, "(") && endsWith(attribute, ")"))
+    {
+        isRegexp = true;
+    }
+    else if(attribute != "")
+    {
+        final_attribute = attribute;
+    }
+    
+    int n = node.children.size();
+    while(n > 0)
+    {
+        isLast = false;
+        RegexpTreeNode child = node.children[n - 1];
+        func(child, regexp_, true, attribute_index, rows, user_agents, set_value, keys_found);
+        --n;
+    }
+    
+    if(isLast)
+    {
+        for (const auto i : collections::range(0, rows))
+        {
+	// Here I want to use match and extractAllGroupsVertical function that can be used by user, but I didn't find where did they come from, so I just left them here for now.
+            if(match(user_agents[i], regexp_) == 1)
+            {
+                if(isRegexp)
+                {
+                    final_attribute = std::regex_replace(attribute, "\\1", extractAllGroupsVertical(user_agents[i], node.regexp)[0][0]);
+                }
+                set_value(i, final_attribute);
+                ++keys_found;
+                break;
+            }
+        }
+    }
+    for(size_t i = 0; i < node.regexp.size(); ++i)
+    {
+	regexp_.pop_back();
```

Comment:
Maybe `regexp_` should be `vector<string>` ? Actualy it's unclear why we concatenate regexps into this variable